Board support for MIMXRT1062-FMURT6

sumitbatra-nxp opened the pull request:
This PR includes code to add board support for mimxrt1062_fmurt6 board -

1. 	Adds code to build and flash on mimxrt1062-fmurt6 board with SoC part MIMXRT1062DVL6B. Refer [https://www.nxp.com/part/RDDRONE-FMURT6#/](url)
2. Adds pinmux and device tree entries for making use of the CAN, Ethernet, I2C, SPI, PWM 
and UART interfaces on mimxrt1062-fmurt6 board

sumitbatra-nxp commented on boards/arm/mimxrt1062_fmurt6/mimxrt1062_fmurt6.yaml:
```diff
@@ -0,0 +1,30 @@
+  - can
+  - watchdog
+  - adc
+  - pwm
```
@iuliana-prodan .. Thanks for your valuable comment.. I have made the change

sumitbatra-nxp commented:
Can someone please review the changes, as all the necessary checks have passed.

danieldegrasse commented on boards/arm/mimxrt1062_fmurt6/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+#
+
+if (CONFIG_DISPLAY)
+message(WARNING "
```
This is a message specific to NXP's RT1050 and 106x EVKs, because these boards may cause damage to their display circuit if it is enabled when no display is attached. Is this something you need for the FMURT6 board?

sumitbatra-nxp replied:
Thanks Daniel.. This is indeed not required for FMURT6.. hence removing it..

kartben replied:
This has in fact not been removed: https://github.com/zephyrproject-rtos/zephyr/blob/5781478fd5c466d0fdab82af96580026b099c24e/boards/arm/mimxrt1062_fmurt6/CMakeLists.txt#L7-L11

danieldegrasse commented on boards/arm/mimxrt1062_fmurt6/Kconfig.defconfig:
```diff
@@ -0,0 +1,34 @@
+	default y if BOARD_MIMXRT1062_FMURT6
+
+config FLASH_MCUX_FLEXSPI_NOR
+	default n if BOARD_MIMXRT1062_FMURT6
```
I don't believe you should need `default n` here, as the driver would not be enabled by default?

sumitbatra-nxp replied:
@danieldegrasse .. This board has a flexspi hyperflash part S26KS512SDPBHI020 so I have made -
FLASH_MCUX_FLEXSPI_HYPERFLASH as default y and 
FLASH_MCUX_FLEXSPI_NOR as default n
![image](https://github.com/zephyrproject-rtos/zephyr/assets/107573027/56ea7eef-81c3-4b00-92a3-69fc899f07a1)

Let me know of your thoughts..

danieldegrasse replied:
It looks like you have removed both in the latest push- is that correct? `CONFIG_FLASH_MCUX_FLEXSPI_HYPERFLASH` should be automatically enabled when `CONFIG_FLASH=y`, based on the compatible of the flash device in your board devicetree

sumitbatra-nxp replied:
Yes, after verifying that CONFIG_FLASH_MCUX_FLEXSPI_HYPERFLASH is enabled based on the compatible field I removed both..

sumitbatra-nxp replied:
Made the changes as suggested

danieldegrasse commented on boards/arm/mimxrt1062_fmurt6/board.cmake:
```diff
@@ -0,0 +1,15 @@
+board_runner_args(pyocd "--target=mimxrt1060")
+board_runner_args(jlink "--device=MIMXRT1062xxx6A")
+
+if (${CONFIG_BOARD_MIMXRT1062_FMURT6})
```
Why guard this with a CONFIG? It seems like this setting will always be required

sumitbatra-nxp replied:
Removed the check as suggested

danieldegrasse commented on boards/arm/mimxrt1062_fmurt6/doc/index.rst:
```diff
@@ -0,0 +1,402 @@
++-----------+------------+-------------------------------------+
+| ADC       | on-chip    | adc                                 |
++-----------+------------+-------------------------------------+
+| SAI       | on-chip    | i2s                                 |
```
Have all these IP blocks been enabled and tested with this board?

sumitbatra-nxp replied:
I have not tested them, as it is not yet required. I can remove them from the table if you suggest.

danieldegrasse replied:
Understood. Please do add any IP blocks that have been tested and verified with this board to this list, I just want to make sure that we don't confuse users if an IP block has not been tested on this board.

sumitbatra-nxp replied:
Added all the relevant IP blocks that have been tested and removed the non tested ones

sumitbatra-nxp replied:
@danieldegrasse .. This section and the connections and IO section has been appropriately updated now

danieldegrasse commented on boards/arm/mimxrt1062_fmurt6/mimxrt1062_fmurt6-pinctrl.dtsi:
```diff
@@ -0,0 +1,522 @@
+				<&iomuxc_gpio_emc_41_gpio3_io27>,
+				<&iomuxc_gpio_ad_b0_00_gpio1_io00>,
+				<&iomuxc_snvs_pmic_on_req_gpio5_io01>;
+			//bias-disable;
```
Please remove this commented line if it is not needed

sumitbatra-nxp replied:
Removed

danieldegrasse commented on boards/arm/mimxrt1062_fmurt6/mimxrt1062_fmurt6_defconfig:
```diff
@@ -0,0 +1,21 @@
+CONFIG_ARM_MPU=y
+CONFIG_HW_STACK_PROTECTION=y
+CONFIG_PINCTRL=y
+CONFIG_FLASH=y
```
Why is `CONFIG_FLASH` enabled by default for this platform? Are flash drivers needed for the "minimal" build- IE running hello world on the board?

sumitbatra-nxp replied:
It was added for an experiment, removed it now.

danieldegrasse commented on soc/arm/nxp_imx/rt/soc_rt10xx.c:
```diff
@@ -54,8 +54,12 @@ const clock_enet_pll_config_t ethPllConfig = {
 	.enableClkOutput500M = true,
 #endif
 #ifdef CONFIG_ETH_MCUX
+#if defined(CONFIG_BOARD_MIMXRT1062_FMURT6)
```
I am unclear why this `#if` is needed. It looks like the ETH clock output will be enabled if `CONFIG_ETH_MCUX` is set

danieldegrasse replied:
Ok, this makes sense. I would rather see us setting this up based on DT nodelabels then. Something like the following:

```
#if DT_NODE_HAS_STATUS(DT_NODELABEL(enet), okay)
enableClkOutput = true,
#endif
#if DT_NODE_HAS_STATUS(DT_NODELABEL(enet2), okay)
enableClkOutput1 = true,
#endif
```

sumitbatra-nxp replied:
This is a valuable comment for me since this clearly marks the difference between Zephyr and Linux, as I think Linux doesn't have DT_HAS type unmodifiable Kconfig options which are based on Device tree.

sumitbatra-nxp replied:
Made the changes as suggested

sumitbatra-nxp replied:
@danieldegrasse Made the changes as suggested

danieldegrasse commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -1021,7 +1022,11 @@ static void eth_mcux_init(const struct device *dev)
 
 #if defined(CONFIG_SOC_SERIES_IMX_RT10XX)
+#if defined(CONFIG_BOARD_MIMXRT1062_FMURT6)
+	sys_clock = CLOCK_GetFreq(kCLOCK_EnetPll1Clk);
```
I'm a bit concerned about introducing this board specific change here. Why does the FMURT board need a different source clock?

sumitbatra-nxp replied:
Such is the board design.
ETH_REF_CLK line between MAC and PHY has only ENET2 in its Pinmux option, such is the board design.

sumitbatra-nxp replied:
Made similar changes here by adding DT_NODE_HAS_STATUS(DT_NODELABEL(enet), okay) and removed than board specific MACRO

danieldegrasse commented on boards/arm/mimxrt1062_fmurt6/doc/index.rst:
```diff
@@ -0,0 +1,402 @@
++---------------+-----------------+---------------------------+
+| WAKEUP        | GPIO            | SW0                       |
++---------------+-----------------+---------------------------+
+| GPIO_B0_00    | LCD_CLK         | LCD Display               |
```
Are all these pins accurate? I'm asking because it does not seem like this board has display support, and this pinmux description should describe the default routing of pins on the board

sumitbatra-nxp replied:
Thanks for pointing out.. I am making precise changes now

sumitbatra-nxp replied:
Made the changes as suggested

sumitbatra-nxp replied:
Changes have been done

mmahadevan108 commented on boards/arm/mimxrt1062_fmurt6/mimxrt1062_fmurt6_defconfig:
```diff
@@ -0,0 +1,21 @@
+
+# Enable Regulators
+CONFIG_REGULATOR=y
+CONFIG_REGULATOR_FIXED_INIT_PRIORITY=75
```
Same comment as above for these regulator `Kconfigs`, why do we need this in the default?

sumitbatra-nxp replied:
I need to enable all the regulators at boot time itself i.e. GPIOs controlling the Enable Line of board's Power sources - VDD_3V3_SD_CARD, VDD_5V_PERIPH, VDD_5V_HIPOWER, reg-3v3-sensor, VDD_3V3_SPEKTRUM, ETH_VDD_3V3.

sumitbatra-nxp commented on boards/arm/mimxrt1062_fmurt6/Kconfig.defconfig:
```diff
@@ -0,0 +1,34 @@
+if FLASH
+
+config FLASH_MCUX_FLEXSPI_HYPERFLASH
+	default y if BOARD_MIMXRT1062_FMURT6
```
Are you suggesting me to remove all the contents of this file ?

sumitbatra-nxp replied:
Made the changes as suggested

sumitbatra-nxp replied:
Changes have been done.. please re-verify

sumitbatra-nxp commented:
@mmahadevan108 @danieldegrasse .. I request you to kindly re-verify.

danieldegrasse commented on soc/arm/nxp_imx/rt/soc_rt10xx.c:
Sorry- one final request here. Can you split the changes to the `soc_rt10xx.c` and `eth_mcux.c` into another commit, separate from the one adding the board?

Something like the following title:

`soc: arm: nxp_imx: support enet2 interface on RT1060 series SOC`

sumitbatra-nxp replied:
Sure.. just to highlight that these changes are particularly for this board.. other RT106x based boards may or may not use these changes.

sumitbatra-nxp replied:
@danieldegrasse .. Thanks for your valuable comment.. I have made this change

danieldegrasse replied:
Thank you for the change- looks good, compliance error just needs to be addressed.

sumitbatra-nxp replied:
All checks are clear now..

sumitbatra-nxp committed:
board: Adds MIMXRT1062-FMURT6 board support package

Also enables all the usable CAN, Ethernet, I2C, PWM,

SPI, UARTs interfaces for this board

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

sumitbatra-nxp committed:
soc: arm: nxp_imx: support enet2 interface on RT106x series

This patch enables the PLL clock output and PLL ref clock
for second ethernet module in NXP's i.MxRT106x SoCs

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

bperseghetti reviewed (approved):
This will be great to get in upstream. Thanks @sumitbatra-nxp @danieldegrasse @mmahadevan108

sumitbatra-nxp commented:
@mmahadevan108.. kindly review as I need 1 more approver :)
